homeassistant/components/cloudflare_r2/__init__.py

```diff
@@ -0,0 +1,87 @@
+"""The Cloudflare R2 integration."""
+
+from __future__ import annotations
+
+import logging
+from typing import cast
+
+from aiobotocore.client import AioBaseClient as S3Client
+from aiobotocore.session import AioSession
+from botocore.exceptions import (
+    ClientError,
+    ConnectionError,
+    EndpointConnectionError,
+    ParamValidationError,
+)
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryError, ConfigEntryNotReady
+
+from .const import (
+    CONF_ACCESS_KEY_ID,
+    CONF_BUCKET,
+    CONF_ENDPOINT_URL,
+    CONF_SECRET_ACCESS_KEY,
+    DATA_BACKUP_AGENT_LISTENERS,
+    DOMAIN,
+)
+
+type R2ConfigEntry = ConfigEntry[S3Client]
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: R2ConfigEntry) -> bool:
+    """Set up Cloudflare R2 from a config entry."""
+
+    data = cast(dict, entry.data)
+    try:
+        session = AioSession()
+        # pylint: disable-next=unnecessary-dunder-call
+        client = await session.create_client(
+            "s3",
+            endpoint_url=data.get(CONF_ENDPOINT_URL),
+            aws_secret_access_key=data[CONF_SECRET_ACCESS_KEY],
+            aws_access_key_id=data[CONF_ACCESS_KEY_ID],
+        ).__aenter__()
```

Comment:
according to #156777 there's a blocking call happening here that we should take care of. I don't have an S3 storage at hand, but something like this should fix it, could you try it out?

```suggestion
        client = hass.async_add_executor_job(
            session.create_client(
                service_name="s3",
                endpoint_url=data.get(CONF_ENDPOINT_URL),
                aws_secret_access_key=data[CONF_SECRET_ACCESS_KEY],
                aws_access_key_id=data[CONF_ACCESS_KEY_ID],
            )
        )
        await client.__aenter__() # pylint: disable-next=unnecessary-dunder-call
```

Reply:
as concluded in #156777 this is a problem with the upstream library, should we leave this as is?